### PR TITLE
Avoid new processes (needs testing)

### PR DESCRIPTION
**Pull request by RoiEXLab**
This PR changes the internal behaviour of TripleA so that the engine doesn't spawn a new JVM just to host or join a game.
Pros:
 - Debuggers work for hosted game scenarios
 - Less overhead because only one JVM is required
 - Avoids having "ghost" java processes running in the background because something crashed

(Current) Cons:
 - A lot of the code call System.exit to terminate a game, this kills of course all other running instances as well, goal would be to find those calls and replace them with something appropriate

I did some testing which included hosting 2 games concurrently (on different ports, because using the same port resulted in an exception as expected), setting them up one after another and joining 2 games at the same time, one after another, because there's just one "selection screen" now.
I haven't gotten any exceptions so far, but I didn't really tried to break anything so far, I just checked if it's possible to host and join multiple games, which works without any severe problems (just an uneccessary popup)

After this PR the only use case for the ProcessUtil class is for the MapCreator classes which are still being invoked using a fresh JVM, and use that as common procedure to invoke other map creator tools :(
So we should probably consider to either move this class to the MapCreator package, or to fix the map creator to not spawn JVMs, which would also fix the LAF issues btw.

So before merging an important goal would be to find any issues with this PR and fix them.

@ssoloff Could you tell me how you set your repo up to create binary builds so we can let the "testing group" do some testing?

### CODE REVIEW

**Comment by ssoloff** [comment]
> Could you tell me how you set your repo up to create binary builds so we can let the "testing group" do some testing?

First, you need to add the Travis environment variables required to do a release build:

* `GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS`
* `GPG_PRIVATE_KEY`
* `GPG_PRIVATE_KEY_PASSPHRASE`
* `INSTALL4J_7_LICENSE`

I think all of those are self-explanatory.  The only one you might not have run into before is the `GPG_PRIVATE_KEY` variable.  That's just a standard GPG-generated private key.  I believe the master secrets document has some footnotes on how to upload that to Travis using either the `travis` command-line tool or the web interface.

The second step is to modify _.travis.yml_ and _.travis/push\_tag_ on your branch.  Here's the last patch I used a few months ago (it may no longer apply cleanly to current `master`):

```diff
diff --git a/.travis.yml b/.travis.yml
index 7973d7e..ba0af6f 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -21,18 +21,15 @@ before_deploy:
 - ./gradlew -PengineVersion="$ENGINE_VERSION" release
 - ./.travis/generate_artifact_checksums
 - ./.travis/push_tag $ENGINE_VERSION
-- ./.travis/push_maps
+#- ./.travis/push_maps
 deploy:
   provider: releases
   api_key:
-    secure: nxaqYrkXLGL3W20/eCnf63DLjMrQAhEuW44jggh1/nI383goa+u6w0bBtWCxRdVzos7t4dpVfS6+kv6oIHacm9zVA+RYrqy5opzCJhq8lmXVVRijbALzUeiFif2HURMaKWj0ynRNVlAyBHzazPTLZVWywifpdSubSkuMWkl20cmuKu/Hg3c1EC9se3OYhhTHx3Hya7xSrctrDEYLsEBAUZzkKfscqRVqwwltS88CgIMtRISDpSBGrtH0t1uAH6NitTSguGgb+QEpqnELcRLymX2G1yzMA4Xr5c/L34MfbBKf8vIuG9t411xYuLoyKoUbroTWxSnPwlSy6PHz+QJ7UCXbDkATOGO3chxlKxglppvI/G3n2YP5Zf2dAaDlHblpvarh55i/4i4sKB2AbvvzkIHrQJwUgmLCbpN8/Vp9GWcGkd6i5U7F8tNInCs6ttX3oGvGOfYEXs02Ctyiea4LAqk4S7GZTuV2QXqxXglL4eRIwZ4UETiwgoAAtHma63Eq7+9t2ykMlk7zAK96FGwJrB97wa08aPuSxL94IYEBmn9Ht/vKXRiNQMvpnfp4rWQtL3cqbVyYAg5EjKb4PsBmnb91+RXtnWFOY1RpZGt8sPXYd+KZYzN1BXTFJEpaLLsIDN6r7nMcAvJDUmucaM+m7giPXz1ZBGAic3UBM1qMCgI=
+    secure: Ln1c9XL66CKtJoM0YpATq8Xh4NsZJ7T4bV2nesGBErtGIuv56F7EejlFBi2N4m8EMv2ulAHyy2N422PebFuF0hvAv+KqGB5+vPCoeuNZ3zYJGMdhUsLOQRw4vSWixfxju8vQltDHHnnOn9Qw+eKDT1IBo+Fuw4C0PlKZHEdkoXHO/4Lkm8EcvCmNJPjRsq9B8UBCioPkbOnJRvm6Vnn4r2v4IniKupdwf+h5IoY91K0l9FC9CiDum/02lNxpDZ620a1Y2SQrYHsY1yr9S5MzXKNYP4aYF5OzbdhuB9IYaVwpe6Ck22W5gnf0d46QBfxIcXUxqMYeIw4L+qAqmkrVXScpBvEIK7V6LFFh9sCxtoY65PFBghu9c+IIB8n1V1kDhyFDEKaCBIHejVmrL0QP1FVT0pHF7eIYQgrLHy4Oy+YAcqyep2YjEYXwC0W36Q1s1SSxlATfIr2uuJaH1KKDaN0okmSejYMvu/xC5m8ONnfISt5P42tR84AbBzNke1HR9VV+DYjme/P3md0rwCQ5iXNNUroDdEb5NJgp+0ssuCluLSK9Th4HIICIyTKaz2vEcgBdAGIfCqQsgYic8AhOBNYfJWmEE3Jheq4RKO8HMlnFq7mWEUyJgU1MoFpN2B8IrS+UYRLx0kYNPDEik9Acwdp8dEtIRAz98GyW0U0TsZg=
   file_glob: true
   file: build/artifacts/*
   skip_cleanup: true
   prerelease: true
   on:
     tags: false
-    repo: triplea-game/triplea
-    branches:
-      only:
-        - master
+    all_branches: true
diff --git a/.travis/push_tag b/.travis/push_tag
index 34d6a62..5524abd 100755
--- a/.travis/push_tag
+++ b/.travis/push_tag
@@ -6,7 +6,7 @@
 # Note, the tag itself is also considered a branch and trigger an extraneous branch build in travis.
 #
 
-TAG_VALUE=$1
+TAG_VALUE="ssoloff-$1"
 
 set -u
 
@@ -14,11 +14,11 @@ git config --global user.email "tripleabuilderbot@gmail.com"
 git config --global user.name "tripleabuilderbot"
 
 echo "Attempting to delete tag $TAG_VALUE, this is so we will recreate the release if this build is being re-run. Ignore any errors from this"
-git push -q https://$GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS@github.com/triplea-game/triplea --delete "refs/tags/$TAG_VALUE"
+git push -q https://$GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS@github.com/ssoloff/triplea-game-triplea --delete "refs/tags/$TAG_VALUE"
 echo "Done deleting existing tag if it existed."
 
 git tag "$TAG_VALUE" -a -f -m "$TAG_VALUE"
 
 echo "Pushing tag: $TAG_VALUE to github to trigger releases deployment"
-git push -q https://$GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS@github.com/triplea-game/triplea --tags
+git push -q https://$GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS@github.com/ssoloff/triplea-game-triplea --tags
 
```

IIRC, the API key specified above for GitHub releases is the same value stored in the `GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS` environment variable but encrypted using the `travis` command-line tool.

<hr>

Let me suggest an alternative idea that may be easier than maintaining a long-running feature branch with releases.  Why not just use the client settings framework to add a new toggle for this feature on the Test tab?  Then this feature can be tested by anyone without having to know of the special binaries in your fork.

File: src/main/java/games/strategy/engine/framework/startup/ui/InGameLobbyWatcher.java [hunk 1]
```diff
@@ -221,20 +221,20 @@ public void connectionAdded(final INode to) {
...
-                + "(The firewall exception must be updated every time a new version of TripleA comes out.)\r\n"
+            final String message = "Your computer is not reachable from the internet.\n"
+                + "Please make sure your Firewall allows incoming connections (hosting) for TripleA.\n"
+                + "(The firewall exception must be updated every time a new version of TripleA comes out.)\n"
                 + "And that your Router is configured to send TCP traffic on port " + portString
                 + " to your local ip address.\r\n"
-                + "See 'How To Host...' in the help menu, at the top of the lobby screen.\r\n"
-                + "The server tried to connect to your external ip: " + addressUsed + "\r\n";
+                + "See 'How To Host...' in the help menu, at the top of the lobby screen.\n"
+                + "The server tried to connect to your external ip: " + addressUsed;
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/startup/ui/InGameLobbyWatcher.java]: The changes in this file appear to be independent of the feature you are adding.  Since this could be a potentially long-running branch, any reason not to just submit them in a separate PR?  They seem useful on their own.

File: src/main/java/games/strategy/engine/framework/GameRunner.java [hunk 2]
```diff
@@ -459,29 +456,25 @@ public static Image getGameIcon(final Window frame) {
...
-    commands.add("-D" + TRIPLEA_SERVER_PROPERTY + "=true");
-    commands.add("-D" + TRIPLEA_PORT_PROPERTY + "=" + port);
-    commands.add("-D" + TRIPLEA_NAME_PROPERTY + "=" + playerName);
-    commands.add("-D" + LOBBY_HOST + "="
-        + messengers.getMessenger().getRemoteServerSocketAddress().getAddress().getHostAddress());
-    commands.add("-D" + GameRunner.TRIPLEA_LOBBY_PORT_PROPERTY + "="
-        + messengers.getMessenger().getRemoteServerSocketAddress().getPort());
-    commands.add("-D" + LOBBY_GAME_COMMENTS + "=" + comments);
-    commands.add("-D" + LOBBY_GAME_HOSTED_BY + "=" + messengers.getMessenger().getLocalNode().getName());
+    System.setProperty(TRIPLEA_SERVER_PROPERTY, String.valueOf(true));
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/GameRunner.java]: Hmm, this seems like it could be a problem.  Do we have to revert the changes to these system properties when what was formerly the "spawned" process ends (e.g. when returning from a hosted game to the main menu)?  It just smells a bit to be changing (possibly multiple times) what are effectively global variables to control the execution path.

File: src/main/java/games/strategy/engine/framework/startup/ui/InGameLobbyWatcher.java (hunk 1 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/startup/ui/InGameLobbyWatcher.java]: Will do

**Comment by RoiEXLab** [comment]
@ssoloff judging by your feedback this PR calls for a bigger refactoring process.
Note to myself: Find usages of the system properties and try to avoid them.

File: src/main/java/games/strategy/engine/framework/GameRunner.java (hunk 2 above)
Comment (ssoloff) [reply src/main/java/games/strategy/engine/framework/GameRunner.java]: Yes, but those changes don't affect the original process that spawned the new process.  Each JVM process gets its own copy of system properties.

**Comment by RoiEXLab** [comment]
@ssoloff Before I proceed:
I thought about making GameRunner non-static and pass a GameRunner instance along with the showMainFrame call, so we can have the exact same behaviour as before.
Any objections?
Might take some days until I get to it, but it'd be better if I knew my work is going in the right direction.

**Comment by ssoloff** [comment]
@RoiEXLab Well, I agree that to get rid of the dependency on the global system properties, _something_ needs to be passed around. :smile:  I'm not sure if that something should be `GameRunner` itself or some kind of configuration object (e.g. `GameRunnerConfiguration`).  You might have to spike this and play around a bit to see what works best.

**Comment by RoiEXLab** [comment]
@ssoloff I made GameRunner non-static which made me discover a lot of unwanted references in packages where this class really doesn't belong (just to display some kind of error message etc.).
I'm not going to fix the checkstyle violations now, because there's still a lot of work to do, but I'd appreciate if you could give me some general feedback about the changes and possibly some suggestions for improvements

**Comment by ssoloff** [comment]
@RoiEXLab Yeah, that really revealed a lot of the mess we have in terms of dependencies, right? :smile:

My first impression is that there is probably a bunch of smaller refactorings that can be done to tease apart the dependencies all these classes have on `GameRunner`, and that those refactorings can probably be done in several smaller PRs.  It's getting late here, so I'll review this in more detail tomorrow to see if I can come up with some ideas.

File: src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java [hunk 3]
```diff
@@ -37,12 +37,13 @@ private BackgroundTaskRunner() {}
    */
   public static void runInBackground(
       final String message,
-      final Runnable backgroundAction)
+      final Runnable backgroundAction,
+      final GameRunner gameRunner)
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java]: `BackgroundTaskRunner` is a good example of a type that probably shouldn't have a dependency on `GameRunner` (and I was the one who introduced it :smile:).  The dependency exists effectively to get a reference to the main frame window.  I think many of the types you had to change in this PR to accept a reference to a `GameRunner` instance are in this category.

A lot of this goes back a few months ago when we removed the `MainFrame` class (and thus the `MainFrame#getInstance()` method) and ended up adding several new UI-related methods to `GameRunner` so the main frame window reference could be encapsulated.  That wasn't necessarily a bad thing, and given what you're trying to achieve in this PR, having a static `MainFrame` instance would have to be addressed anyway.

However, I think this PR will require us to take a step back and re-think how a lot of this global state is managed.  One thing that's been bugging me is how `GameRunner` is some kind of _God Class_.  `GameRunner` is really an orchestration layer for the entire application that brings together many different types of services.  Those services should be extracted into interfaces, and instances of those interfaces passed to the downstream classes instead of `GameRunner` itself.

But let's start out with the low-hanging fruit like this class.  I'll submit a PR that refactors this class a bit so it doesn't depend on `GameRunner` (but its clients still will).  That may spark some further discussion on how to proceed.

File: src/main/java/games/strategy/engine/ClientContext.java [hunk 4]
```diff
@@ -46,7 +47,7 @@
   private static final ClientContext instance = new ClientContext();
 
   private final GameEnginePropertyReader gameEnginePropertyReader = new GameEnginePropertyReader();
-  private final MapDownloadController mapDownloadController = new MapDownloadController();
+  private final MapDownloadController mapDownloadController = new MapDownloadController(new GameRunner());
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/ClientContext.java]: This may indicate that a reference to the `MapDownloadController` should be acquired from `GameRunner` instead of `ClientContext` (i.e. add a `GameRunner::getMapDownloadController()` method).  There should only be a few (ideally one) place where a `GameRunner` instance is created.  And in this case, using a fresh `GameRunner` will mean the Download Maps window displayed by `MapDownloadController` is not associated with the main frame window for the "real" `GameRunner`.

Another way to think about this is that each `GameRunner` may need its own `ClientContext`. 🤔   But that may be going too far.  I think the purpose of `ClientContext` is to provide true singleton instances that shouldn't be specific to any one `GameRunner`.

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/ClientContext.java]: Yes, that was just a dirty hack to avoid the compiler error, couldn't find a better way to pass Amy instance without changing this class completely.

File: src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java (hunk 3 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java]: @ssoloff Assuming you're not going to build on top of this PR, should I just revert the last commit for now to avoid merge conflicts?
Otherwise if you want to build on top of this PR, go ahead and close it. I'd be happy to help you with it.

Comment (ssoloff) [reply src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java]: Sorry, didn't see this comment before I submitted the PR.  Yeah, I didn't think the merge conflicts would be too big a deal (probably mostly changing `GameRunner` to `gameRunner`).

I'm thinking a lot of the changes to break dependencies should be done in `master` and then merged into this branch.  They are probably needed in general, and it will keep this PR from growing into a several-thousand line monstrosity. :smile:

**Comment by RoiEXLab** [comment]
@ssoloff Rebased against current master, more changes will follow. Feel free to tell me if you dislike a specific change

**Comment by RoiEXLab** [comment]
@ssoloff Also if you spot something else that definitely needs to be changed please tell me as well...
My current goal is to be able to make the GameRunner constructor private, and to change it to require the system params afterwards

**Comment by RoiEXLab** [comment]
@ssoloff I'm going to need some help here, the diff is getting way to big:
I created a GameRunnerContext class which basically wraps a HashMap and supposed to be required by the GameRunner constructor.
It's supposed to replace 
```regex
System(\.getProperties\(\))?\.[sg]etProperty\(
```
But there are about 150 matches of this Regex in the codebase, which need their own instance of this context somehow, which is supplied to them somehow

Additionally there are some occurences where a GameRunner instance couldn't be simply passed over because they were enums etc. so we need solutions for that as well.

Any Ideas what could be done to avoid such a large diff (that get's worse with every commit to master)?

**Comment by ssoloff** [comment]
@RoiEXLab Ideally, we'd want to get this change (or something close to it) into `master` so you don't have to keep worrying about this branch diverging too much.  Since `GameRunner` is still static on `master`, could you simply have a static method on that class return the `GameRunnerContext`?  And then replace usage of `System#get`/`setProperty()` using that singleton instance?  To make things easier, you could simply delegate to `System#get`/`setProperty()` within `GameRunnerContext`.  Basically, it would add a level of indirection so you can take baby steps to remove the dependency on system properties.

Slightly off-topic...  Is the setter going to be necessary on `GameRunnerContext`?  Keeping things immutable would be preferable.  If you can't do that, then you may have to do a bit more analysis to determine if the methods of this class could be called from multiple threads.  If so, you'll need to make that aspect of `GameRunnerContext` thread safe (a `ConcurrentHashMap` would probably be sufficient).

**Comment by RoiEXLab** [comment]
@ssoloff I think I get what you're requesting, but I'm unsure how exactly you mean it. Could you elaborate on that or even submit some changes you think of directly to this branch?
Ideally we would push the first 2 commits to a new branch on this repo and work on top of that?

> Slightly off-topic... Is the setter going to be necessary on GameRunnerContext? Keeping things immutable would be preferable. If you can't do that, then you may have to do a bit more analysis to determine if the methods of this class could be called from multiple threads. If so, you'll need to make that aspect of GameRunnerContext thread safe (a ConcurrentHashMap would probably be sufficient)

Not sure, haven't checked, probably not, I simply wanted to mimic the behaviour of System.getProperties, but I just noticed I didn't implement a clearProperty method so whatever...

**Comment by ssoloff** [comment]
@RoiEXLab Yeah, I guess it's hard to explain in prose.  I'll try to code up something this weekend that demonstrates what I'm talking about.

**Comment by ssoloff** [comment]
@RoiEXLab I didn't get to coding up something this weekend, but I plan to try and start it on Monday.

**Comment by ssoloff** [comment]
@RoiEXLab Ok, I started implementing what I suggested above.  While working on it, I started to get a better picture of what we're up against here.  You've already identified the problem of how to pass a `GameRunnerContext` around to all the places where it would be needed.  In addition, here are a few other things I observed related to breaking the dependency on system properties:

1. We use both Java-managed and TripleA-managed system properties.  The Java-managed properties have nothing to do with what we're trying to encapsulate within `GameRunnerContext`.  It appears someone realized this in the past and introduced the `g.s.engine.framework.system.SystemProperties` class.  I think this is a good idea, and **I will submit a PR to `master` to move all TripleA usage of Java-managed system properties to this class**.  That will allow us to focus on exactly what we need to.
1. `GameRunner` isn't the only TripleA entry point.  We have to consider `HeadlessGameServer` and `LobbyServer` (to a lesser extent), as well.  Each of these entry points will be responsible for managing a context object that will flow down appropriately.  That suggests the `GameRunner` prefix on the context class may be too specific.  That is, there probably only needs to be a single type that can be used by all three entry points; I'm not sure we'll need `GameRunnerContext`, `HeadlessGameServerContext`, and `LobbyServerContext` types.  Or if we do have those three concrete types, they may all have to implement a common interface because there is downstream code that can be called in both `GameRunner` and `HeadlessGameServer` processes.
1. Instead of trying to capture all the existing system properties into the context at once, it might be easier to move them incrementally one at a time.  I felt kind of overwhelmed trying to follow how the information flows in several of the more complex cases, so starting with the low-hanging fruit may help us build up confidence.

**Comment by RoiEXLab** [comment]
@ssoloff closing this PR then.
I will keep the branch around
